### DISCUSSION TITLE
Generating Typescript Interfaces

### DISCUSSION THREAD

**Discussion by gpollo** [wasm-bindgen/wasm-bindgen#3254] [Q&A]
Hey everybody, does anybody know if we can generate Typescript `interface` using `wasm-bindgen`?

For example, let's say I have this Rust code:

```rust
#[wasm_bindgen]
pub struct MyInterface {
    pub a: i32,
    pub b: Option<f32>,
}
```

It will currently generate this Typescript code:

```typescript
export class TestTest {
    free(): void;
    a: number;
    b?: number;
}
```

I'm wondering if we could generate a `interface` instead of `class`:

```typescript
export interface TestTest {
    a: number;
    b?: number;
}
```

While `class` and `interface` are similar in Typescript, they have different meaning and it might be ideal to use one instead of the other.

**Comment by haraldreingruber-dedalus** [comment]
What about exporting a Rust `trait` as a Typescript interface?
The use case might be a bit different, and I know traits and interfaces are not completely the same thing, but not sure if the difference is relevant...

My use case would be to export 2 different structs/classes, that share a common interface (trait from Rust perspective), and I would like to treat them the same on the TS side in most of the situations, but still have knowledge about their specific type...

Not sure if that could be easily implemented, but I guess in the worst case by passing parameters like in the following example could make it not too difficult?

```Rust
#[wasm_bindgen]
trait Vehicle {
    fn speed(&self) -> f64;
}

#[wasm_bindgen(implements_trait = "Vehicle")]
struct Car {
    max_speed: f64,
}

impl Vehicle for Car {
    fn speed(&self) -> f64 {
        self.max_speed
    }
}

#[wasm_bindgen(implements_trait = "Vehicle")]
struct Bicycle {
    max_speed: f64,
}

impl Vehicle for Bicycle {
    fn speed(&self) -> f64 {
        self.max_speed
    }
}
```

Or is there already a workaround to achieve this today, by customizing the wasm_bindgen code generation?

This issue looks like a proposal in a similar direction:
https://github.com/rustwasm/wasm-bindgen/issues/2073
and a first Draft PR:
https://github.com/rustwasm/wasm-bindgen/pull/2871

**Comment by ImUrX** [comment]
That's why we use Symbol for the interface names, because a Rust trait method can have the same name as a struct's method :3

**Comment by sisou** [comment]
https://github.com/madonoharu/tsify generates `interface` from structs.
